File: 366-find-leaves-of-binary-tree/366-find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        def recurse(current: TreeNode) -> int:
            if not current:
                return 0
            
            i = max(recurse(current.left), recurse(current.right))
            if len(ans) < (i + 1):
                ans.append([])
            ans[i].append(current.val)

            return i + 1
                
        ans = []
        recurse(root)
        return ans

File: 366-find-leaves-of-binary-tree/366-find-leaves-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        this.helper(root, ans);
        return ans;
    }
    
    private int helper(TreeNode curr, List<List<Integer>> ans){
        if(curr == null){
            return 0;
        }
        
        int left = this.helper(curr.left, ans);
        int right = this.helper(curr.right, ans);
        
        int i = Math.max(left, right);
        if(i > ans.size() - 1){
            ans.add(new ArrayList<>());
        }
        
        ans.get(i).add(curr.val);
        
        return i + 1;
    }
}